### PR TITLE
Save image link

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

_Reviewer: @leoi11_

Acceptance criteria:
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.

Dependencies:
- Deployment Environment

### PR DESCRIPTION
**Pull request by blakeherrington**
Closes #2 

Seems to be working as the ticket describes. Let me know your comments! Thanks so much!

- [ x] I assigned the designated reviewer to this PR

### CODE REVIEW

**Comment by blakeherrington** [comment]
Mmm... gonna look at resolving the conflicts in the branch.

**Comment by blakeherrington** [comment]
All ready for review 👍

**Comment by rmacklin** [comment]
@blakeherrington FYI, there's no need to keep the "I assigned the designated reviewer to this PR" line instead of actually checking it off. It's mainly just there as a reminder when you're opening the PR.

File: Gemfile [hunk 1]
```diff
@@ -21,6 +21,9 @@ gem 'turbolinks'
 gem 'jbuilder', '~> 2.0'
 # bundle exec rake doc:rails generates the API under doc/api.
 gem 'sdoc', '~> 0.4.0', group: :doc
+# gem Simple form
+gem 'simple_form'
+# bootstrap
```
Comment (rmacklin) [inline Gemfile]: I'm curious about this comment.

File: Gemfile [hunk 2]
```diff
@@ -21,6 +21,9 @@ gem 'turbolinks'
 gem 'jbuilder', '~> 2.0'
 # bundle exec rake doc:rails generates the API under doc/api.
 gem 'sdoc', '~> 0.4.0', group: :doc
+# gem Simple form
+gem 'simple_form'
```
Comment (rmacklin) [inline Gemfile]: Minor, but I'd add this to the bottom in alphabetical order, rather than putting it up here in the middle of the generated Gemfile content. Also, I don't think the comment is necessary (it doesn't provide additional value).

File: Gemfile.lock [hunk 3]
```diff
@@ -72,23 +72,25 @@ GEM
     jbuilder (2.4.1)
       activesupport (>= 3.0.0, < 5.1)
       multi_json (~> 1.2)
-    jquery-rails (4.1.0)
-      rails-dom-testing (~> 1.0)
+    jquery-rails (4.1.1)
+      rails-dom-testing (>= 1, < 3)
```
Comment (rmacklin) [inline Gemfile.lock]: This is strange. There are a bunch of changes to the `Gemfile.lock` that seem unrelated to the addition of `simple_form`. Try reverting the changes in the `Gemfile.lock` and then rereunning `bundle install`. It shouldn't change unrelated gems.

File: app/views/images/index.html.erb [hunk 4]
```diff
@@ -1 +1,6 @@
 <h1 id="hello_world"><%= @hello_world %></h1>
+
+<%= simple_form_for @image do |f| %>
+    <%= f.input :url %>
+    <%= f.button :submit %>
+<% end %>
```
Comment (rmacklin) [inline app/views/images/index.html.erb]: The index page will be used to show all the added images. As such, it'd be preferable to move this "new image form" to a different page: the images/new page.

Connor originally did this same thing, and I made the same comment. He has addressed it, so he may be able to help you if you're not sure what I mean.

File: db/migrate/20160405222756_create_images.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class CreateImages < ActiveRecord::Migration
+  def change
+    create_table :images do |t|
+      t.string :url
+
+      t.timestamps null: false
```
Comment (rmacklin) [inline db/migrate/20160405222756_create_images.rb]: 👍 Nice job remembering to add the timestamps.

File: app/controllers/images_controller.rb [hunk 6]
```diff
@@ -1,6 +1,18 @@
 # Basic images controller
 class ImagesController < ApplicationController
   def index
-    @hello_world = 'Hello World!'
+    @hello_world = "Hello World!"
+    @image = Image.new
+  end
+
+  def create
+    @image = Image.create(create_params)
```
Comment (rmacklin) [inline app/controllers/images_controller.rb]: You should use `create!` here so that it will raise an exception (instead of just returning `false`) if the creation fails.

File: test/fixtures/images.yml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
+  url: MyString
+
+two:
+  url: MyString
```
Comment (rmacklin) [inline test/fixtures/images.yml]: We generally don't want to use fixtures (they end up causing more trouble than benefit), so I'd delete this file.

File: test/models/image_test.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
+  # end
+end
```
Comment (rmacklin) [inline test/models/image_test.rb]: There's no need to commit this when it's empty.

**Comment by rmacklin** [comment]
The review app appears to be broken. It's probably because you need to manually run the migrations you added. You should check the logs to verify that:
```
heroku logs --app blake-image-sharer-stage-pr-19
```
and then after you've verified it, run:
```
heroku run rake db:migrate --app blake-image-sharer-stage-pr-19
```

**Comment by blakeherrington** [comment]
Closing to try to regenerate review app

**Comment by rmacklin** [comment]
@blakeherrington Looks good. In the future, please try to make [atomic commits](https://sites.google.com/a/appfolio.com/eng/resources/our-tools/git). In this case, it would have been appropriate to add `simple_form` (and the related config) in one commit, and implement the save image feature in a second commit.

File: app/controllers/images_controller.rb [hunk 9]
```diff
@@ -1,6 +1,21 @@
 # Basic images controller
 class ImagesController < ApplicationController
   def index
-    @hello_world = 'Hello World!'
+    @hello_world = "Hello World!"
```
Comment (rmacklin) [inline app/controllers/images_controller.rb]: I'm surprised to see this change from single quotes back to double quotes.

File: app/models/image.rb [hunk 10]
```diff
@@ -0,0 +1,3 @@
+class Image < ActiveRecord::Base
+  validates_presence_of :url
```
Comment (rmacklin) [inline app/models/image.rb]: Validating the presence of the `url` attribute was not actually part of this story, but since you did it there should be a corresponding test for this.

File: app/views/images/_form.html.erb [hunk 11]
```diff
@@ -0,0 +1,4 @@
+<%= simple_form_for @image do |f| %>
+    <%= f.input :url %>
+    <%= f.button :submit %>
+<% end %>
```
Comment (rmacklin) [inline app/views/images/_form.html.erb]: Your indentation seems to be off here too. If that's RubyMine's doing, you should be able to configure the indentation to 2 spaces in the Preferences. As for your earlier question, pronto didnt complain because it's only running against `.rb` files.

File: app/views/images/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,2 @@
+<%= link_to 'Home', root_path %>
+<%= render partial: 'form', locals: { image: @image } %>
```
Comment (rmacklin) [inline app/views/images/new.html.erb]: Since you're passing this as a local, you should refer to the local variable within that partial (on [this line](https://github.com/AppFolioOnboarding/image-sharer-blakeherrington/pull/19/files#diff-34fc81efe413385c2c036d3eca4c1a52R1)), not the instance variable.

File: test/controllers/images_controller_test.rb [hunk 13]
```diff
@@ -6,4 +6,32 @@ class ImagesControllerTest < ActionController::TestCase
     assert_response :success
     assert_select '#hello_world', 1
   end
+
+  test 'create' do
+    test_image_url = 'http://test.com/example_image.png'
+    params = { image: { url: 'http://test.com/example_image.png' } }
```
Comment (rmacklin) [inline test/controllers/images_controller_test.rb]: Did you mean to set this to `test_image_url`?

File: test/controllers/images_controller_test.rb [hunk 14]
```diff
@@ -6,4 +6,32 @@ class ImagesControllerTest < ActionController::TestCase
...
+
+    assert_difference 'Image.count', 1 do
+      post :create, params
+    end
+    assert_response :success
+
+    image = Image.last
+    assert_equal image.url, test_image_url
```
Comment (rmacklin) [inline test/controllers/images_controller_test.rb]: The order of `assert_equal` is "expected" then "actual". This matters for when the test fails, because it will say something like "expected foo, got bar", which is confusing if the order is wrong.

File: db/schema.rb [hunk 15]
```diff
@@ -1,18 +1,22 @@
...
 # system, you should be using db:schema:load, not running all the migrations
-# from scratch. The latter is a flawed and unsustainable approach (the more
-# migrations you'll amass, the slower it'll run and the greater likelihood
-# for issues).
+# from scratch. The latter is a flawed and unsustainable approach (the more migrations
+# you'll amass, the slower it'll run and the greater likelihood for issues).
 #
-# It's strongly recommended that you check this file into your version control
-# system.
+# It's strongly recommended that you check this file into your version control system.
```
Comment (rmacklin) [inline db/schema.rb]: As [above](https://github.com/AppFolioOnboarding/image-sharer-blakeherrington/pull/19/files#r58791159), I'm surprised to see these changes.
